icu4c/source/test/intltest/loctest.cpp

```diff
@@ -3154,3 +3155,117 @@ void LocaleTest::TestBug13417VeryLongLanguageTag() {
     status.errIfFailureAndReset("\"%s\"", l.getName());
     assertEquals("equals", tag, result.c_str());
 }
+
+void LocaleTest::TestBug11053UnderlineTimeZone() {
+    static const char* const tz_in_ext[] = {
+        "etadd",
+        "tzdar",
+        "eheai",
+        "sttms",
+        "arirj",
+        "arrgl",
+        "aruaq",
+        "arluq",
+        "mxpvr",
+        "brbvb",
+        "arbue",
+        "caycb",
+        "brcgr",
+        "cayzs",
+        "crsjo",
+        "caydq",
+        "svsal",
+        "cafne",
+        "caglb",
+        "cagoo",
+        "tcgdt",
+        "ustel",
+        "bolpb",
+        "uslax",
+        "sxphi",
+        "mxmex",
+        "usnyc",
+        "usxul",
+        "usndcnt",
+        "usndnsl",
+        "ttpos",
+        "brpvh",
+        "prsju",
+        "clpuq",
+        "caffs",
+        "cayek",
+        "brrbr",
+        "mxstis",
+        "dosdq",
+        "brsao",
+        "gpsbh",
+        "casjf",
+        "knbas",
+        "lccas",
+        "vistt",
+        "vcsvd",
+        "cayyn",
+        "cathu",
+        "hkhkg",
+        "mykul",
+        "khpnh",
+        "cvrai",
+        "gsgrv",
+        "shshn",
+        "aubhq",
+        "auldh",
+        "imdgs",
+        "smsai",
+        "asppg",
+        "pgpom",
+    };
+    static const char* const tzname_with_underline[] = {
+        "America/Buenos_Aires",
+        "America/Coral_Harbour",
+        "America/Los_Angeles",
+        "America/Mexico_City",
+        "America/New_York",
+        "America/Rio_Branco",
+        "America/Sao_Paulo",
+        "America/St_Johns",
+        "America/St_Thomas",
+        "Australia/Broken_Hill",
+        "Australia/Lord_Howe",
+        "Pacific/Pago_Pago"
+    };
+    std::string locale_str;
+    for (int32_t i = 0; i < UPRV_LENGTHOF(tz_in_ext); i++) {
+      locale_str = "en-u-tz-";
+      locale_str += tz_in_ext[i];
+      Locale l(locale_str.c_str());
+      assertTrue((locale_str + " !l.isBogus()").c_str(), !l.isBogus());
+    }
+    for (int32_t i = 0; i < UPRV_LENGTHOF(tzname_with_underline); i++) {
+      locale_str = "en@timezone=";
+      locale_str +=  tzname_with_underline[i];
+      Locale l(locale_str.c_str());
+      assertTrue((locale_str + " !l.isBogus()").c_str(), !l.isBogus());
+
+      locale_str = "en_US@timezone=";
+      locale_str +=  tzname_with_underline[i];
+      Locale l2(locale_str.c_str());
+      assertTrue((locale_str + " !l2.isBogus()").c_str(), !l2.isBogus());
+
+      locale_str = "en_Latn@timezone=";
+      locale_str +=  tzname_with_underline[i];
+      Locale l3(locale_str.c_str());
+      assertTrue((locale_str + " !l3.isBogus()").c_str(), !l3.isBogus());
+
+      locale_str = "en_Latn_US@timezone=";
+      locale_str +=  tzname_with_underline[i];
```

Comment:
I'm sceptical that it adds any real value (and not just more runtime) to test each and every one of these 48 combinations of base locales and timezones. Is there any conceivable situation in which one of them could fail while the others pass?

Have you considered updating the `tzname_with_underline` list to contain full locale IDs, that won't need to be concatenated with anything, and just having a small well-chosen set of different test cases instead of testing all of these combinations?